File: apps/athenaeum/native/AthenaeumDomain/Sources/AthenaeumDomain/ChangesMessage.swift
import Foundation

private enum ChangesMessageValidationError: Error {
    case emptyTitle
    case nonPositiveVersion
}

// Mirrors `packages/domain/src/changes-message.ts` — the `changes` stream envelope (plan:
// "Acceptance rides the same changes-message stream, gaining createdNodes/addedFacts/addedEdges/
// noteEdits fields; mergeChanges/revertChanges promote/delete pending records exactly as
// multi-gadget.md §Q15 describes"). Each batch field is a lightweight summary record, not the full
// entity — see changes-message.ts's header comment for why.

/// Mirrors `changes-message.ts`'s `CreatedNodeSummary`: `{nodeId, title}`.
public struct CreatedNodeSummary: Codable, Hashable, Sendable {
    public let nodeId: EntityId
    public let title: String

    public init(nodeId: EntityId, title: String) {
        self.nodeId = nodeId
        self.title = title
    }
}

/// Mirrors `changes-message.ts`'s `AddedFactSummary`: `{factId, nodeId, predicateId}`.
public struct AddedFactSummary: Codable, Hashable, Sendable {
    public let factId: EntityId
    public let nodeId: EntityId
    public let predicateId: String

    public init(factId: EntityId, nodeId: EntityId, predicateId: String) {
        self.factId = factId
        self.nodeId = nodeId
        self.predicateId = predicateId
    }
}

/// Mirrors `changes-message.ts`'s `AddedEdgeSummary`: `{edgeId, relationDefinitionId,
/// sourceNodeId, targetNodeId}`.
public struct AddedEdgeSummary: Codable, Hashable, Sendable {
    public let edgeId: EntityId
    public let relationDefinitionId: EntityId
    public let sourceNodeId: EntityId
    public let targetNodeId: EntityId

    public init(edgeId: EntityId, relationDefinitionId: EntityId, sourceNodeId: EntityId, targetNodeId: EntityId) {
        self.edgeId = edgeId
        self.relationDefinitionId = relationDefinitionId
        self.sourceNodeId = sourceNodeId
        self.targetNodeId = targetNodeId
    }
}

/// Mirrors `changes-message.ts`'s `NoteEditSummary`: `{nodeId, headsHash}` — a note-body edit
/// accepted via the Automerge-fork mechanism (`ChatForkService`, not this file's `pending`-flag
/// batch fields — see changes-message.ts's doc comment).
public struct NoteEditSummary: Codable, Hashable, Sendable {
    public let nodeId: EntityId
    public let headsHash: String

    public init(nodeId: EntityId, headsHash: String) {
        self.nodeId = nodeId
        self.headsHash = headsHash
    }
}

/// Mirrors `changes-message.ts`'s `CreatedAppSummary`: `{appId, title}`.
public struct CreatedAppSummary: Codable, Hashable, Sendable {
    public let appId: EntityId
    public let title: String

    public init(appId: EntityId, title: String) throws {
        guard !title.isEmpty else {
            throw ChangesMessageValidationError.emptyTitle
        }
        self.appId = appId
        self.title = title
    }

    private enum CodingKeys: String, CodingKey { case appId, title }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        try self.init(
            appId: container.decode(EntityId.self, forKey: .appId),
            title: container.decode(String.self, forKey: .title)
        )
    }
}

/// Mirrors `app.ts`'s `AppCodeKind = Schema.Literal("client", "server")`.
public enum AppCodeKind: String, Codable, Hashable, Sendable {
    case client
    case server
}

/// Mirrors `changes-message.ts`'s `UpdatedAppCodeSummary`: `{appId, kind, version}`.
public struct UpdatedAppCodeSummary: Codable, Hashable, Sendable {
    public let appId: EntityId
    public let kind: AppCodeKind
    public let version: Int

    public init(appId: EntityId, kind: AppCodeKind, version: Int) throws {
        guard version > 0 else {
            throw ChangesMessageValidationError.nonPositiveVersion
        }
        self.appId = appId
        self.kind = kind
        self.version = version
    }

    private enum CodingKeys: String, CodingKey { case appId, kind, version }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        try self.init(
            appId: container.decode(EntityId.self, forKey: .appId),
            kind: container.decode(AppCodeKind.self, forKey: .kind),
            version: container.decode(Int.self, forKey: .version)
        )
    }
}

/// Mirrors `changes-message.ts`'s `ChangesMessage`: `{chatId, sequence, createdNodes?,
/// addedFacts?, addedEdges?, noteEdits?}` — the `changes` stream envelope. All four batch fields
/// are independently optional (per §Q15's "a creation-only batch has an empty no-op update"),
/// including the app-library extensions `createdApps` and `updatedAppCode`.
public struct ChangesMessage: Codable, Hashable, Sendable {
    public let chatId: EntityId
    public let sequence: Int
    public let createdNodes: [CreatedNodeSummary]?
    public let addedFacts: [AddedFactSummary]?
    public let addedEdges: [AddedEdgeSummary]?
    public let noteEdits: [NoteEditSummary]?
    public let createdApps: [CreatedAppSummary]?
    public let updatedAppCode: [UpdatedAppCodeSummary]?

    public init(
        chatId: EntityId,
        sequence: Int,
        createdNodes: [CreatedNodeSummary]? = nil,
        addedFacts: [AddedFactSummary]? = nil,
        addedEdges: [AddedEdgeSummary]? = nil,
        noteEdits: [NoteEditSummary]? = nil,
        createdApps: [CreatedAppSummary]? = nil,
        updatedAppCode: [UpdatedAppCodeSummary]? = nil
    ) {
        self.chatId = chatId
        self.sequence = sequence
        self.createdNodes = createdNodes
        self.addedFacts = addedFacts
        self.addedEdges = addedEdges
        self.noteEdits = noteEdits
        self.createdApps = createdApps
        self.updatedAppCode = updatedAppCode
    }
}

File: apps/athenaeum/packages/domain/src/changes-message.ts
import * as Schema from "effect/Schema"
import { AppCodeKind } from "./app.js"
import { EntityId } from "./node.js"

// Phase 3 storage-schema task (plan: "Acceptance rides the same changes-message stream, gaining
// createdNodes/addedFacts/addedEdges/noteEdits fields; mergeChanges/revertChanges promote/delete
// pending records exactly as multi-gadget.md §Q15 describes"). `ChangesMessage` is the
// `changes`-collection envelope named in the plan's §"Storage & domain model" collection list
// ("chats, changes — agent-edit provisional-change stream").
//
// Element-shape decision: each of the four batch fields below is a lightweight *summary* record
// (an id plus just enough display context), not the full `Node`/`Fact`/`Edge` entity. This
// mirrors `multi-gadget.md` §Q15's own `createdGadgets: {gadgetId, title}[]` field precisely —
// the `changes` message's job is bookkeeping (which entities did this batch touch, for
// crash-recovery set-difference matching and for the accept/revert UI's summary line), not being
// the source of truth for the entities themselves. The full pending record already lives in its
// own collection (`nodes`/`facts`/`edges`, each carrying `pending: {chatId, sequence?}` — see
// node.ts/fact.ts/edge.ts); duplicating the whole row into the stream as well would just be a
// second, driftable copy of the same data.

/** One node created by this batch — mirrors `multi-gadget.md`'s `createdGadgets` element shape
 *  (`{gadgetId, title}`) field-for-field, substituting `nodeId` for `gadgetId`. */
export class CreatedNodeSummary extends Schema.Class<CreatedNodeSummary>("CreatedNodeSummary")({
  nodeId: EntityId,
  title: Schema.String.pipe(Schema.minLength(1))
}) {}

/** One fact added by this batch — enough context (`nodeId`, `predicateId`) to render "added
 *  `predicateId` on `nodeId`" in an accept/revert summary without a second full `Fact` copy. */
export class AddedFactSummary extends Schema.Class<AddedFactSummary>("AddedFactSummary")({
  factId: EntityId,
  nodeId: EntityId,
  predicateId: Schema.String.pipe(Schema.minLength(1))
}) {}

/** One edge added by this batch — enough context to render "linked `sourceNodeId` →
 *  `targetNodeId`" without a second full `Edge` copy. */
export class AddedEdgeSummary extends Schema.Class<AddedEdgeSummary>("AddedEdgeSummary")({
  edgeId: EntityId,
  relationDefinitionId: EntityId,
  sourceNodeId: EntityId,
  targetNodeId: EntityId
}) {}

/**
 * One note-body edit accepted by this batch. Note-body edits use the Automerge-fork mechanism
 * (see chat-fork-rpc.ts / docs/automerge-fork-spike.md), not the row-level `pending` flag, so
 * there is no pending `Page`/`Node` row for a `noteEdits` entry to summarize the way
 * `CreatedNodeSummary` et al. do — `headsHash` (mirroring `Page.headsHash`, see page.ts) is what
 * a reader needs to know a fork's accept actually landed and which mainline doc state resulted,
 * without re-deriving it from the Automerge document itself.
 */
export class NoteEditSummary extends Schema.Class<NoteEditSummary>("NoteEditSummary")({
  nodeId: EntityId,
  headsHash: Schema.String.pipe(Schema.minLength(1))
}) {}

/**
 * One App created by this batch — App Library domain-extension task addition, mirroring
 * `CreatedNodeSummary`'s own `{nodeId, title}` shape field-for-field, substituting `appId` for
 * `nodeId`. The full pending `App` row (with its own `pending` marker) already lives in the
 * `App` entity itself (app.ts); this is bookkeeping only, same rationale as this file's header
 * comment gives for the other three summary kinds.
 */
export class CreatedAppSummary extends Schema.Class<CreatedAppSummary>("CreatedAppSummary")({
  appId: EntityId,
  title: Schema.String.pipe(Schema.minLength(1))
}) {}

/**
 * One App code update accepted by this batch — App Library domain-extension task addition. Unlike
 * `CreatedAppSummary` (which mirrors the other three "new pending row" summaries), a proposed code
 * update does not create a second pending `App`/`AppCodeVersion` *record type* to summarize the
 * existence of — it creates an ahead-of-pointer `AppCodeVersion` row (see app.ts's
 * `AppCodeVersion` doc comment for the full mechanism). `kind`/`version` here are exactly what a
 * `mergeChanges`/`revertChanges` implementation (agent-edit-rpc.ts) needs to know which pointer
 * (`App.clientCodeVersion` or `App.serverCodeVersion`) to advance, or which ahead-of-pointer row
 * to delete, without re-deriving it from the chat's raw tool-call log.
 */
export class UpdatedAppCodeSummary extends Schema.Class<UpdatedAppCodeSummary>(
  "UpdatedAppCodeSummary"
)({
  appId: EntityId,
  kind: AppCodeKind,
  version: Schema.Number.pipe(Schema.int(), Schema.greaterThan(0))
}) {}

/**
 * The `changes` stream envelope. Mirrors the plan's exact field list: `{chatId, sequence,
 * createdNodes?, addedFacts?, addedEdges?, noteEdits?}`, widened by the App Library
 * domain-extension task with two further optional batch fields (`createdApps`, `updatedAppCode`)
 * following the identical extension pattern — see those two summary classes' own doc comments.
 *
 * `sequence` is the batch's own position in the chat's changes stream — the same number
 * `PendingMarker.sequence` (node.ts) is stamped with for every pending record this message
 * covers, and the value `mergeChanges(chatId, mergeThrough)`/`revertChanges(chatId, revertFrom)`
 * compare against (plan §Q15: "mergeChanges... promotes... creations with sequence <=
 * mergeThrough"). All six batch fields are optional and independently omittable — per §Q15's
 * "a creation-only batch has an empty no-op update," a single `changes` message may carry any
 * non-empty subset of them (e.g. a turn that only calls `addFact` produces a message with
 * `addedFacts` set and the rest absent).
 */
export class ChangesMessage extends Schema.Class<ChangesMessage>("ChangesMessage")({
  chatId: EntityId,
  sequence: Schema.Number.pipe(Schema.int(), Schema.nonNegative()),
  createdNodes: Schema.optional(Schema.Array(CreatedNodeSummary)),
  addedFacts: Schema.optional(Schema.Array(AddedFactSummary)),
  addedEdges: Schema.optional(Schema.Array(AddedEdgeSummary)),
  noteEdits: Schema.optional(Schema.Array(NoteEditSummary)),
  createdApps: Schema.optional(Schema.Array(CreatedAppSummary)),
  updatedAppCode: Schema.optional(Schema.Array(UpdatedAppCodeSummary))
}) {}
